django/contrib/admin/views/main.py

```diff
@@ -380,3 +387,140 @@ def url_for_result(self, result):
                                                self.opts.model_name),
                        args=(quote(pk),),
                        current_app=self.model_admin.admin_site.name)
+
+
+class AutocompletePage(Page):
+    def has_next(self):
+        """
+        Return true until queryset is smaller than the requested page size.
+
+        This avoids calling calculating the whole page count.
+        """
+        return self.object_list.count() >= self.paginator.per_page
+
+
+class AutocompletePaginator(Paginator):
+    """
+    A Django Paginator for big data sets.
+
+    Because users mostly use the first pages, it doesn't matter
+    if the page count is correct or not, as long as it's high enough.
+
+    Count() queries can take time, because they want to be exact.
+
+    So this paginator just doesn't count.
+    """
+
+    def _get_page(self, *args, **kwargs):
+        return AutocompletePage(*args, **kwargs)
+
+    def validate_number(self, number):
+        """Validates the given 1-based page number but doesn't care about empty pages."""
+        try:
+            number = int(number)
+        except (TypeError, ValueError):
+            raise PageNotAnInteger('That page number is not an integer')
+        if number < 1:
+            raise EmptyPage('That page number is less than 1')
+        return number
+
+    @cached_property
+    def count(self):
+        if self.object_list.exists():
+            return sys.maxsize
+        return 0
+
+
+class AutocompleteJsonView(BaseListView):
+    """
+    View that handles requests from `.AutocompleteWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginator_class = AutocompletePaginator
+    paginate_by = 20
+    admin_site = None
+
+    @property
+    def model(self):
+        return self.field.remote_field.model
+
+    @property
+    def field(self):
+        from django.apps import apps
+        try:
+            original_model = apps.get_model(self.app_label, self.model_name)
+            model_admin = self.admin_site._registry[original_model]
+            if not model_admin.has_change_permission(self.request):
+                # The user does not have access to see this field.
+                raise PermissionDenied
+            return original_model._meta.get_field(self.field_name)
+        except (LookupError, FieldDoesNotExist):
+            raise PermissionDenied('Invalid "field_identifier".')
+
+    @property
+    def model_admin(self):
+        try:
+            return self.admin_site._registry[self.model]
+        except KeyError:
+            raise PermissionDenied('%s is not registered in the admin.' % self.model)
```

Comment:
Doesn't this reveal information about what exists and what doesn't? I'd rather just use `Raise PermissionDenied`